lms/resources/_js_config.py

```diff
@@ -40,9 +40,12 @@ def add_canvas_file_id(self, canvas_file_id):
         :raise HTTPBadRequest: if a request param needed to generate the config
             is missing
         """
-        self._config["api"]["viaCallbackUrl"] = self._request.route_url(
-            "canvas_api.files.via_url", file_id=canvas_file_id
-        )
+        self._config["api"]["viaUrl"] = {
+            "authUrl": self._request.route_url("canvas_api.authorize"),
+            "path": self._request.route_path(
+                "canvas_api.files.via_url", file_id=canvas_file_id
+            ),
```

Comment:
All this stuff looks Canvas specific. Should it name-spaced accordingly? `api.viaUrl` seems a bit general.

It also seems like it might be a bit redunant. The paths we have here are:

 * `api.viaUrl.path`
 * `api.viaUrl.authUrl`

Might something like this be better?

* `api.canvas.via.url`
* `api.canvas.via.authUrl`

Reply:
I think the redundancy might be confusion:

* The `viaUrl` in `api.viaUrl` refers to the ["Via URL" API](https://github.com/hypothesis/lms/blob/626fecfe5942df22cb5291de1ec22472cedd2a53/lms/views/api/canvas/files.py#L23-L39), so named because it's an API that returns a `{ via_url: <VIA_URL> }` response where `<VIA_URL>` is a URL suitable for use as the `src` of the assignment's Via iframe. Currently the Via URL endpoint handles only Canvas files. In future there'll either be a second Via URL endpoint for Blackboard files or the single endpoint will be capable of handling either
* The `path` in `api.viaUrl.path` is the path _to the Via URL API_, for example `/api/canvas/files/23/via_url`. In future this will sometimes be the path to a Blackboard Via URL API instead. All API endpoints in the JS config have a `path` like this

So there's no redundancy -- the `viaUrl` and `path` refer to different things. `viaUrl` is the name of the API endpoint, `path` is the path to that endpoint. `path` is also just a path not a URL.

`viaUrl` actually _isn't_ Canvas specific. The way it works is:

* If the Via URL can be known up front then the backend puts a `config.viaUrl` setting in the JS config. Currently happens for public URL and Google Drive assignments in any LMS
* If the Via URL can't be known up front then the backend doesn't put a `config.viaUrl` setting in the JS config, but does put a `config.api.viaUrl` object instead. The frontend has to call this API to get the Via URL. Happens for Canvas files assignments and soon for Blackboard files assignments. This is because we need to call the Canvas or Blackboard API to get a temporary public URL for the file on every launch

So `config.api.viaUrl` will either contain a Canvas `authUrl` and `path` or a Blackboard `authUrl` and `path` depending. But this is opaque to the frontend. The frontend knows only `config.api.viaUrl` and needn't know whether it's Canvas or Blackboard. The backend just handles setting it appropriately and the frontend responds.

There can never be more than one `viaUrl` in play at once, so there's no need to support having two of them in different locations in the config (and doing so would complicate both the backend and frontend code)

Reply:
Ok, so what's in there in a Canvas context is Canvas specific, but in different context you get different content in the same place? Kind of make sense.

I notice that the contents are totally different depending on whether this is set by `add_document_url()` or `add_canvas_file_id()`. I presume this is because the former is actually the URL to go to, and the later is a callback saying "go here to find out"?

If that's the case it makes sense, but some short comments to that effect would probably be good. It's not obvious. e.g.:

```
# Tell the frontend to call us back for a URL
# Give the frontend an exact URL to use for the document
```

Also `canvasOAuth2RedirectError.authorizeUrl?`

Reply:
Also AFAIK, a path is just a non-fully qualified URL, so if we wanted to use the name it's accurate.

Reply:
Added a couple of comments to those docstrings.

It's important to distinguish between "paths" and "URLs" here because various things in the JS config are just paths and the JS prepends the host to them, whereas other things are full URLs. We should probably fix this so that the backend always sends full URLs and the frontend doesn't prepend anything, at some point.

Reply:
It's kind of an unrelated change so to reduce risk I've sent a separate PR to rename `authorizeUrl`, so that can be reviewed and tested separately: https://github.com/hypothesis/lms/pull/2092

Reply:
Yep that all seems very sensible. The the front-end mucks about with it we need to mark it as one or the other. I'll take a look at the other PR.